Sources/ClientRuntime/Networking/Streaming/BufferedStream.swift

```diff
@@ -0,0 +1,128 @@
+/// The buffer size depends on the amount of data written and read.
+/// Note: This class is thread-safe.
+/// Note: if data is not read from the stream, the buffer will grow indefinitely until the stream is closed.
+///       or reach the maximum size of a `Data` object.
+public class BufferedStream: Stream {
+
+    /// Returns the length of the stream, if known
+    public private(set) var length: Int?
+
+    /// Returns the current position in the stream
+    public private(set) var position: Data.Index
+
+    /// Returns true if the stream is empty, false otherwise
+    public var isEmpty: Bool {
+        return buffer.isEmpty
+    }
+
+    /// Returns true if the stream is seekable, false otherwise
+    /// Note: This stream is seekable, but seeking to a position that is not in the buffer will cause the stream to throw an error.
+    public var isSeekable: Bool {
+        return false
+    }
+
+    private var isClosed: Bool
+
+    private(set) var buffer = Data()
+    private let lock = NSRecursiveLock()
+
+    /// Initializes a new `BufferedStream` instance.
+    /// - Parameters:
+    ///   - data: The initial data to buffer.
+    ///   - isClosed: Whether the stream is closed.
+    public init(data: Data = .init(), isClosed: Bool = false) {
+        self.buffer = data
+        self.position = data.startIndex
+        self.length = data.count
+        self.isClosed = isClosed
+    }
+
+    /// Reads up to `count` bytes from the stream.
+    /// - Parameter count: The maximum number of bytes to read.
+    /// - Returns: Data read from the stream, or nil if the stream is closed and no data is available.
+    public func read(upToCount count: Int) throws -> Data? {
+        lock.withLockingClosure {
+            let toRead = min(count, buffer.count)
+            let endPosition = position.advanced(by: toRead)
+            let chunk = buffer[position..<endPosition]
+
+            // remove the data we just read
+            buffer.removeFirst(toRead)
+
+            // update position
+            position = endPosition
+
+            // if we're closed and there's no data left, return nil
+            // this will signal the end of the stream
+            if isClosed && chunk.isEmpty {
+                return nil
+            }
+
+            return chunk
+        }
+    }
+
+    /// Reads all data from the stream.
+    /// It uses `read(upToCount:)` to read data in chunks.
+    /// - Returns: Data read from the stream, or nil if the stream is closed and no data is available.
+    public func readToEnd() throws -> Data? {
+        var data = Data()
+
+        while let next = try read(upToCount: Int.max) {
+            data.append(next)
+        }
+
+        return lock.withLockingClosure {
+            // if we're closed and there's no data left, return nil
+            // this will signal the end of the stream
+            if isClosed && data.isEmpty {
+                return nil
+            }
+
+            return data
+        }
```

Comment:
Now that we have the recursive lock, let's add the entire contents in the locking closure 

```suggestion
        return lock.withLockingClosure {
            var data = Data()

            while let next = try read(upToCount: Int.max) {
                data.append(next)
            }
            
            // if we're closed and there's no data left, return nil
            // this will signal the end of the stream
            if isClosed && data.isEmpty {
                return nil
            }

            return data
        }
```

Reply:
that would block the read or write, depending what comes first.

Here the producer and consumer are 1:1. And we never expect multiple but they can be on different threads.


`readToEnd` `write` `readToEnd` `write` `write` `readToEnd`

something like this.

Reply:
I don't think I understand. My suggested change is needed to prevent race conditions, where the stream may be written to from Thread A, while Thread B is in the middle of reading to end. Without making my suggested change, there is no need for a recursive lock.

Reply:
talked offline, updated it per our discussion.